### PR TITLE
Task/apps 2471 tud data download dashboard

### PR DESCRIPTION
**Pull request by anzioka**
PR:
A basic dashboard for retrieving time use diary submissions constrained by a date range, and creating a csv file for all submissions by day. 

The new Time Use Diary tab is only visible for studies that are associated with the Time Use Diary survey entity.

Screen Shot 2020-11-10 at 4 21 36 PM

no submissions for specified date range
Screen Shot 2020-11-10 at 4 21 23 PM

csv file snapshots:
Screen Shot 2020-11-10 at 2 18 05 PM

Screen Shot 2020-11-10 at 2 17 04 PM

### CODE REVIEW

File: src/containers/studies/StudiesSagas.js [hunk 1]
```diff
@@ -133,6 +136,53 @@ const { ENROLLED, NOT_ENROLLED } = EnrollmentStatuses;
...
+        }
+      });
+    });
+
+    yield put(getTimeUseDiaryStudies.success(action.id, timeUseDiaryStudies));
+  }
+  catch (error) {
+    workerResponse.error = error;
+    LOG.error(action.type, error);
+  }
```
Comment (UnsungHero97) [inline src/containers/studies/StudiesSagas.js:183]: Missing `finally` block.

File: src/containers/tud/TimeUseDiarySagas.js [hunk 2]
```diff
@@ -50,8 +108,247 @@ function* submitTudDataWatcher() :Saga<*> {
...
+    const submissionIds = entities.map((entity) => entity.get(OPENLATTICE_ID_FQN)).flatten();
+    const submissionMetadata = Map(entities.map((entity) => [entity.getIn([OPENLATTICE_ID_FQN, 0]), entity]));
+
+    // entity set ids
+    const submissionESID = yield select(selectEntitySetId(SUBMISSION_ES_NAME));
+    const answersESID = yield select(selectEntitySetId(ANSWERS_ES_NAME));
+    const registeredForESID = yield select(selectEntitySetId(REGISTERED_FOR_ES));
+    const questionsESID = yield select(selectEntitySetId(QUESTIONS_ES_NAME));
+    const timeRangeESID = yield select(selectEntitySetId(TIMERANGE_ES_NAME));
+    const addressesESID = yield select(selectEntitySetId(ADDRESSES_ES_NAME));
```
Comment (UnsungHero97) [inline src/containers/tud/TimeUseDiarySagas.js:222]: No change required. Depending on the structure of your redux store, you could use `selectEntitySets` from `lattice-utils`.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

**Review by UnsungHero97 — DISMISSED** [review]

File: src/containers/studies/StudiesReducer.js [hunk 3]
```diff
@@ -285,6 +290,16 @@ export default function studiesReducer(state :Map<*, *> = INITIAL_STATE, action
...
     }
 
+    case getTimeUseDiaryStudies.case(action.type): {
+      return getTimeUseDiaryStudies.reducer(state, action, {
+        REQUEST: () => state.setIn([GET_TIME_USE_DIARY_STUDIES, REQUEST_STATE], RequestStates.PENDING),
+        FAILURE: () => state.setIn([GET_TIME_USE_DIARY_STUDIES, REQUEST_STATE], RequestStates.FAILURE),
+        SUCCESS: () => state
+          .setIn([GET_TIME_USE_DIARY_STUDIES, REQUEST_STATE], RequestStates.SUCCESS)
+          .set(TIME_USE_DIARY_STUDIES, action.value)
+      });
```
Comment (bergieboy) [inline src/containers/studies/StudiesReducer.js:303]: You added `finally` block to Saga, but not here. So should we add:
```
.setIn([GET_TIME_USE_DIARY_STUDIES, action.id], action)
```
to `REQUEST`, and:
```
.deleteIn([GET_TIME_USE_DIARY_STUDIES, action.id])
```
to `FINALLY`?

File: src/containers/studies/StudiesSagas.js [hunk 4]
```diff
@@ -773,13 +826,20 @@ function* getStudiesWorker(action :SequenceAction) :Generator<*, *, *> {
...
     if (!authorizedStudies.isEmpty()) {
+      // get notification status for authorized studies
       response = yield call(getStudyNotificationStatusWorker, getStudyNotificationStatus({
         studiesEntitySetId,
         studies: authorizedStudies
       }));
       if (response.error) throw response.error;
+
+      // get studies that have time use diary
+      const studyEntityKeyIds = authorizedStudies.map((study) => study.getIn([OPENLATTICE_ID_FQN, 0]));
```
Comment (bergieboy) [inline src/containers/studies/StudiesSagas.js]: use `getEntityKeyId` from `lattice-uitils`

File: src/containers/study/StudyDetailsContainer.js [hunk 5]
```diff
@@ -113,6 +127,9 @@ const StudyDetailsContainer = (props :Props) => {
         <Route
             path={Routes.QUESTIONNAIRES}
             render={() => <QuestionnairesContainer study={study} />} />
+        <Route
+            path={Routes.TUD_DASHBOARD}
+            render={() => <TimeUseDiaryDashboard studyEKID={getEntityKeyId(study)} studyId={studyUUID} />} />
```
Comment (bergieboy) [inline src/containers/study/StudyDetailsContainer.js]: `studyEKID` is already defined above. No need to `getEntityKeyId` twice.

File: src/containers/tud/TimeUseDiaryReducer.js [hunk 6]
```diff
@@ -1,30 +1,69 @@
...
+      });
+    }
+
+    case downloadTudResponses.case(action.type): {
+      const date = action.value;
+      return downloadTudResponses.reducer(state, action, {
+        REQUEST: () => state.setIn([DOWNLOAD_TUD_RESPONSES, REQUEST_STATE, date], RequestStates.PENDING),
+        FAILURE: () => state.setIn([DOWNLOAD_TUD_RESPONSES, REQUEST_STATE, date], RequestStates.FAILURE),
+        SUCCESS: () => state.setIn([DOWNLOAD_TUD_RESPONSES, REQUEST_STATE, date], RequestStates.SUCCESS)
+      });
```
Comment (bergieboy) [inline src/containers/tud/TimeUseDiaryReducer.js]: `finally` blocks are in sagas, but no `FINALLY` cases here.

File: src/containers/tud/TimeUseDiarySagas.js [hunk 7]
```diff
@@ -3,22 +3,80 @@
...
+  PERSON_ID,
+  VALUES_FQN,
+} = PROPERTY_TYPE_FQNS;
+
+const {
+  QUESTIONS_ES_NAME,
+  ANSWERS_ES_NAME,
+  TIMERANGE_ES_NAME,
+  SUBMISSION_ES_NAME
+} = ENTITY_SET_NAMES;
```
Comment (bergieboy) [inline src/containers/tud/TimeUseDiarySagas.js]: sort

File: src/containers/tud/TimeUseDiarySagas.js [hunk 8]
```diff
@@ -50,8 +108,247 @@ function* submitTudDataWatcher() :Saga<*> {
...
+        propertyTypeIds: [personPTID]
+      }
+    ));
+    if (participantsRes.error) throw participantsRes.error;
+    const participants :Object = participantsRes.data.reduce((result, entity) => ({
+      [getIn(entity, [OPENLATTICE_ID_FQN, 0])]: getIn(entity, [PERSON_ID, 0]),
+      ...result
+    }), {});
+
+    const participantEKIDs :UUID[] = participantsRes.data.map((entity) => getEntityKeyId(entity));
```
Comment (bergieboy) [inline src/containers/tud/TimeUseDiarySagas.js]:
```
const participantEKIDs :UUID[] = participantsRes.data.map(getEntityKeyId);
```

File: src/containers/tud/components/SearchPanel.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import styled from 'styled-components';
+
+const { isPending } = ReduxUtils;
+
+const SearchGrid = styled.div`
+  display: grid;
+  grid-template-columns: auto auto 1fr 200px;
+  grid-gap: 30px;
+  align-items: end;
+`;
```
Comment (bergieboy) [inline src/containers/tud/components/SearchPanel.js]: sort

File: src/containers/tud/components/SearchPanel.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+import { ReduxUtils } from 'lattice-utils';
+
+import type { RequestState } from 'redux-reqseq';
+
+import {
+  Typography,
+  Button,
+  Label,
+  DatePicker
```
Comment (bergieboy) [inline src/containers/tud/components/SearchPanel.js:14]: sort imports

File: src/containers/tud/components/SummaryListComponent.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+const SummaryListComponent = (
+  {
+    date,
+    entities,
+    onDownloadData,
+    downloadRS
+  } :Props
+) => (
```
Comment (bergieboy) [inline src/containers/tud/components/SummaryListComponent.js]: sort

**Review by bergieboy — CHANGES_REQUESTED** [review]

File: src/containers/tud/TimeUseDiarySagas.js (hunk 8 above)
Comment (anzioka) [reply src/containers/tud/TimeUseDiarySagas.js]: awesome!

File: src/containers/studies/StudiesReducer.js (hunk 3 above)
Comment (anzioka) [reply src/containers/studies/StudiesReducer.js:303]: Agreed.

**Review by bergieboy — APPROVED** [review]

**Review by UnsungHero97 — APPROVED** [review]

**Pull request merged**